### PR TITLE
Fix backup post hook issue

### LINKED ISSUE #8159: Post hook command is executed before the Kopia snapshot is completed
**What steps did you take and what happened:**

I am using Velero to back up a PostgreSQL dump created by a pre-hook command, so that Kopia can backup the dump, then delete the file using a post-hook command. However, I noticed that the post-hook was executed before Kopia had the chance to back up the dump.

I conducted several tests:
- The first one involved adding a `sleep 60` at the beginning of my post-hook command-> the dump was successfully included in the Kopia snapshot.

- The second test involved creating a file after the dump was completed, which I used in my post-hook command to wait for the file to be fully generated before deleting the dump -> the dump was not included in the Kopia snapshot.

**What did you expect to happen:**

Dump file should be in Kopia snapshot

**Anything else you would like to add:**

```yaml
apiVersion: velero.io/v1
kind: Schedule
metadata:
  name: my-app-pg-app
  namespace: velero
spec:
  schedule: "0 4 * * *"
  useOwnerReferencesInBackup: false
  template:
    includedNamespaces:
      - my-app
    labelSelector:
      matchLabels:
        cnpg.io/cluster: pg-app-16
    metadata: {}
    ttl: 720h
    hooks:
      resources:
        - name: dump
          pre:
            - exec:
                container: postgres
                command:
                  - /bin/bash
                  - -c
                  - |
                    set -e
                    echo "/pgdata/*" > /var/lib/postgresql/data/.kopiaignore &&
                    pg_dump -d app -f /var/lib/postgresql/data/dump_app-$(hostname)-$(date '+%Y-%m-%d-%H-%M-%S')
                onError: Fail
                timeout: 5m
          post:
            - exec:
                container: postgres
                command:
                  - /bin/bash
                  - -c
                  - |
                    rm -f /var/lib/postgresql/data/dump_app-$(hostname)-*
                onError: Fail

```

**Environment:**

- Velero version (use `velero version`):  1.14.0
- Velero features (use `velero client config get features`):  None
- Kubernetes version (use `kubectl version`): 1.30
- Kubernetes installer & version: -
- Cloud provider or hardware configuration: AWS - EKS v1.30.2-eks-db838b0
- OS (e.g. from `/etc/os-release`): Amazon Linux 2023

**Vote on this issue!**

This is an invitation to the Velero community to vote on issues, you can see the project's [top voted issues listed here](https://github.com/vmware-tanzu/velero/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).  
Use the "reaction smiley face" up to the right of this comment to vote.

- :+1: for "I would like to see this bug fixed as soon as possible"
- :-1: for "There are more important bugs to focus on right now"

### PR DESCRIPTION
**Pull request by ywk253100**
Fixes https://github.com/vmware-tanzu/velero/issues/8159 on `main` branch.

This PR introduces an overall `Handler` to handle all kinds of hooks.

The basic workflow is as follows:
1. The `Parser` parses the hooks from different specs into a common `ResourceHook` structure.
2. The `Handler` handles the `ResourceHook` by delegating to a specific `ResourceHandler` according to the hook type:
    - `podBackupPreHookHandler`: handles the backup pre hook for pods
    - `podBackupPostHookHandler`: handles the backup post hook for pods

The changes in this PR will be cherry-picked to branch 1.15 to fix #8159.

And will create another PR to cover restore init/exec hooks for pods and the PR will be only available on `main` (tracked by #8507 )

### CODE REVIEW

**Comment by kaovilai** [comment]
changelog check

File: changelogs/unreleased/8490-ywk253100 [hunk 1]
```diff
@@ -0,0 +1 @@
+Fix backup post hook issue
```
Comment (kaovilai) [inline changelogs/unreleased/8490-ywk253100:1]:
```suggestion
Fix Backup post hooks sometimes executed before PVBs are handled after #7571
```

File: internal/hook/handler.go [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+	return results
+}
+
+func (h *handler) AsyncHandleResourceHooks(ctx context.Context, log logrus.FieldLogger, resource *unstructured.Unstructured, hooks []*ResourceHook) {
+	n := len(hooks)
+	h.WaitGroup.Add(n)
+	go func() {
+		defer func() {
+			for i := 0; i < n; i++ {
+				h.Done()
```
Comment (kaovilai) [inline internal/hook/handler.go:136]:
```suggestion
				h.Done() // Done decrements the [WaitGroup](https://pkg.go.dev/sync#WaitGroup) counter by one.
```

Comment (ywk253100) [reply internal/hook/handler.go]: Updated.

File: changelogs/unreleased/8490-ywk253100 (hunk 1 above)
Comment (ywk253100) [reply changelogs/unreleased/8490-ywk253100]: Updated

File: internal/hook/pod_backup_post_hook_handler.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+// Don't need to wait when using CSI snapshot, because:
+//
+//	PVCs are backed up before Pods (Pod's BIA returns PVCs as additional resources).
+//	PVC's BIA creates VS and returns it as additional resource.
+//	VS's BIA waits VSC's handle ready: https://github.com/vmware-tanzu/velero/blob/v1.14.1/pkg/backup/actions/csi/volumesnapshot_action.go#L118
+//
+// So when Pods are backed up (when hooks execute), the snapshot is taken already
+func (p *podBackupPostHookHandler) WaitUntilReadyToExec(ctx context.Context, log logrus.FieldLogger, res *unstructured.Unstructured) error {
+	// wait all PVBs processed
+	_, err := p.podVolumeBackupper.WaitAllPodVolumesProcessed(log)
```
Comment (ywk253100) [inline internal/hook/pod_backup_post_hook_handler.go:41]: Here is a limitation:
The logic here waits all PVBs of the whole backup to be completed before executing the hooks. This delays the hook execution of one pod although it isn't necessary.

The ideal behavior should be only wait the PVBs belongs to one ItemBlock to be completed before executing the hooks, but this involves the refactor for podvolume.Backupper to make is capable to check whether the PVBs of one ItemBlock completed. The effort isn't tiny, so created another issue #8506 to track.

**Comment by ywk253100** [comment]
Created a new PR https://github.com/vmware-tanzu/velero/pull/8509 to fix the issue, closing this one.

**Pull request closed without merging**